gpu-compute: Added functions to choose replacement policies for GPU

yuxiaojia opened the pull request:
Adding RP_choose functions to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement 
policies  for TCC, TCP and SQC caches for GPU

yuxiaojia committed:
Adding functions to choose replacement policies for GPU

Adding RP_choose functions to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies  for TCC, TCP and SQC caches for GPU

yuxiaojia committed:
Adding functions to choose replacement policies for GPU

Adding RP_choose functions to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies  for TCC, TCP and SQC caches for GPU

yuxiaojia committed:
gpu-compute: Added functions to choose replacement policies for GPU

Adding RP_choose function to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies  for TCC, TCP and SQC caches for GPU

mattsinc commented:
@yuxiaojia : you will need to update the commit message to use the appropriate format and have a ChangeID before this can be merged in

See https://www.gem5.org/contributing "Contributing" for details.

mattsinc reviewed (commented):
In terms of code, this seems fine -- but like I said you won't be able to merge until you fix commit message and add a ChangeID

yuxiaojia committed:
gpu-compute: Added functions to choose replacement policies for GPU

Adding RP_choose functions to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies for TCC, TCP and SQC caches for GPU

yuxiaojia committed:
gpu-compute: Added functions to choose replacement policies for GPU

Adding RP_choose functions to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies for TCC, TCP and SQC caches for GPU

Change-Id: If84a13babf1006ad41a557747c45d48ce2ce22a9

yuxiaojia commented:
> In terms of code, this seems fine -- but like I said you won't be able to merge until you fix commit message and add a ChangeID

I will try to fix it right now

yuxiaojia committed:
gpu-compute: Added functions to choose replacement policies for GPU

Adding RP_choose function to change replacement policies among
TreePLRU, LRU, FIFO, LFU, LIP, MRU, NRU, RRIP, SecondChance AND ShiPMem replacement
policies  for TCC, TCP and SQC caches for GPU

Change-Id: I86cc41cca19f8e0d24d8cf015e2e034a1fc4bc43

yuxiaojia committed:
Change black format

Change-Id: I3733b31baf187e0d3d38d971d9423a1b1afe2296

mattsinc reviewed (approved):
@abmerop would be great if you can check to, but maybe not needed

abmerop commented:
You need to change the GPUFS file too.  This will (did) break GPUFS.

mattsinc commented:
Thanks @abmerop -- @yuxiaojia can you please make the equivalent changes there?

mattsinc commented:
Shoot, I just realize that I merged this already -- @yuxiaojia please make a separate commit updating the GPUFS file then.  Do this as your first priority since it is breaking GPUFS.

abmerop commented:
I put a commit here: https://github.com/gem5/gem5/pull/1230

odanrc commented on configs/example/apu_se.py:
```diff
@@ -380,6 +380,28 @@
+    "--tcp-rp",
+    type=str,
+    default="TreePLRURP",
+    help="cache replacement policy" "policy for tcp",
```
I am very late to the party, but this is how the PR should have been implemented to avoid hardcoding:

Add this line to each of these new arguments:
        choices=ObjectList.rp_list.get_names(),

odanrc commented on configs/ruby/GPU_VIPER.py:
```diff
@@ -149,7 +149,7 @@ def create(self, options):
         self.assoc = options.tcp_assoc
         self.resourceStalls = options.no_tcc_resource_stalls
-        self.replacement_policy = TreePLRURP()
+        self.replacement_policy = RP_choose(options.tcp_rp)
```
Replace these lines using RP_choose by:
self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp)

odanrc commented on configs/ruby/GPU_VIPER.py:
```diff
@@ -1124,3 +1124,28 @@ def create_system(
     return (cpu_sequencers, dir_cntrl_nodes, mainCluster)
+
+
+def RP_choose(test_name):
```
This is a hardcoded version of gem5/configs/ObjectList.py::rp_list

yuxiaojia replied:
> This is a hardcoded version of gem5/configs/ObjectList.py::rp_list

Hi, from what my understanding now, the change does not work since, for example, if we want to make self.replacement_policy = TreePLRURP(), however the self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp) will not make the same change to call the class TreePLRURP(), since it would generate an error that "could not create sytem for ruby protocol GPU_VIPER". Please let me know if it makes sense, or I understand the get() function incorrectly.

odanrc replied:
I may not have understood your question. Are you saying that you wanted to use ObjectList.rp_list.get() to replace the assignments ´self.replacement_policy = TreePLRURP()´? If so, then you would probably have to do it like ´ObjectList.rp_list.get("TreePLRURP")´, IIRC, but I have never tried doing that. Regardless, this line is not the major focus of my suggestion for deduplication, so it does not need to be changed to use the object list. Are you having trouble making the line ´self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp)´ work?

yuxiaojia replied:
Yes, I am not able to make the line  ´self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp)´  work now since it will not create the ruby system as error said, and that TreePLRURP is an example I had

odanrc replied:
Could you give me a command line to run a test case so that I can try creating the PR?

yuxiaojia replied:
Sorry for the late response. You could run an example command line to both use RRIP for TCP and TCC:

`docker run --volume $(pwd):$(pwd) -w $(pwd) docker.io/library/gccimage build/VEGA_X86/gem5.opt  configs/example/apu_se.py -n 3 --dgpu --gfx-version=gfx900 --num-compute-units=4 --cu-per-sa=4 --num-gpu-complex=1 --reg-alloc-policy=dynamic --num-tccs=8 --num-dirs=64 --mem-size=16GB --mem-type=HBM_2000_4H_1x64 --vreg-file-size=16384 --sreg-file-size=800 --tcc-size=4MB --gpu-clock=1801MHz --ruby-clock=1000MHz --vrf_lm_bus_latency=6 --mem-req-latency=69 --mem-resp-latency=69 --mandatory_queue_latency=1 --max-cu-tokens=160 --max-coalesces-per-cycle=10 --sqc-size=16kB --tcp-size=4MB --scalar-mem-req-latency=28 --TCP_latency=4 --tcp-assoc=16 --tcp-num-banks=16 --TCC_latency=121 --tcc-assoc=16 --tcc-tag-access-latency=1 --tcc-data-access-latency=1 --tcp-rp=RRIPRP --WB_L2 --tcc-rp=RRIPRP -c TESTNAME`

The docker.io/library/gccimage can be built by 
```
cd util/dockerfiles/gcn-gpu/
docker build -t gccimage .
```

odanrc replied:
Thanks for the command. I could not run the docker image at all (could not find shared object for python3.10), but when I ran the command manually I found the issue.

Where I said to put:
 self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp)
Should have been (the last 2 characters were missing):
 self.replacement_policy = ObjectList.rp_list.get(options.tcp_rp)()

I created a PR to fix this. In theory I have tested it, but I hadn´t worked in gem5 since gerrit, so if you could double check that I tested this appropriately, it wold be great: https://github.com/gem5/gem5/pull/1530

mattsinc replied:
Thanks @odanrc !  I'm not sure which docker @yuxiaojia is using, but in theory you should be able to test yours using the same instructions here: https://github.com/gem5/gem5-resources/blob/develop/src/gpu/square/README.md

odanrc commented:
I can´t really run gem5 in my machine to test a commit, so I won´t risk breaking stuff uploading a PR with the changes I propose, but if you can easily test this here are my suggestions on how to make a PR to improve this code by removing code duplication and hardcoded strings.

mattsinc commented:
@yuxiaojia : can you please make a new commit that incorporates @odanrc 's suggestions (and of course test them first)?

yuxiaojia commented:
> @yuxiaojia : can you please make a new commit that incorporates @odanrc 's suggestions (and of course test them first)?

Hi Profess Sinclair, I tried to make the change as suggested but seems like it caused an error that "could not create sytem for ruby protocol GPU_VIPER." I am currently looking into this problem.

mattsinc commented:
> > @yuxiaojia : can you please make a new commit that incorporates @odanrc 's suggestions (and of course test them first)?
> 
> Hi Profess Sinclair, I tried to make the change as suggested but seems like it caused an error that "could not create sytem for ruby protocol GPU_VIPER." I am currently looking into this problem.

Ok, feel free to message me or post here if you need help.
